hotspot/src/share/vm/opto/type.cpp

```diff
@@ -2553,8 +2553,11 @@ TypeOopPtr::TypeOopPtr(TYPES t, PTR ptr, ciKlass* k, bool xk, ciObject* o, int o
                    _offset >= InstanceMirrorKlass::offset_of_static_fields()) {
           // Static fields
           assert(o != NULL, "must be constant");
-          ciInstanceKlass* k = o->as_instance()->java_lang_Class_klass()->as_instance_klass();
-          ciField* field = k->get_field_by_offset(_offset, true);
+          ciField* field = NULL;
+          if (o != NULL) {
+            ciInstanceKlass* k = o->as_instance()->java_lang_Class_klass()->as_instance_klass();
+            field = k->get_field_by_offset(_offset, true);
+          }
```

Comment:
This is useless papering over, not a fix, and not even good because it makes the code less legible at no gain.

Better to convert the two `assert`s to actual checks which abort the control flow of the function if triggered.

Reply:
(disclaimer: I am not a reviewer)
This proposed change doesn't really handle the case where o == NULL, although it makes the compiler happy.  I propose something like:

    assert(o != NULL, "must be constant");
    if (o != NULL) {
        ciInstanceKlass* k = o->as_instance()->java_lang_Class_klass()->as_instance_klass();
        ciField* field = k->get_field_by_offset(_offset, true);
        assert(field != NULL, "missing field");
        BasicType basic_elem_type = field->layout_type();
        _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
                                                     basic_elem_type == T_ARRAY);
     } else {
        _is_ptr_to_narrowoop = (?? I am not familiar enough with this code to put a suggestion here)
    }

Reply:
No, don’t `assert` *and* `if` something. (The compiler may delete one of these two under Undefined Behaviour rules.)

I’ve not looked at the callers, but returning an error or throwing an exception if `o == NULL` would be best. Do **not** rely on `assert`.

Reply:
This is a good suggestion, I will try to fix it, refer to https://github.com/openjdk/jdk/blob/ad78b7fa67ba30cab2e8f496e4c765be15deeca6/src/hotspot/share/opto/type.cpp#L3560C3-L3572

thanks.